src/cursor.rs

```diff
@@ -0,0 +1,116 @@
+use std::ops::{Add, AddAssign, Index, IndexMut, Sub, SubAssign};
+
+// TODO: AddAssign
+pub struct CursorMut<'a, T> {
+    data: &'a mut [T],
+    index: usize,
+}
+
+impl<'a, T> CursorMut<'a, T> {
+    pub fn new(data: &'a mut [T]) -> Self {
+        CursorMut { data, index: 0 }
+    }
+
+    pub fn as_slice(&self) -> &[T] {
+        &self.data[self.index..]
+    }
+
+    pub fn as_mut_slice(&mut self) -> &mut [T] {
+        &mut self.data[self.index..]
+    }
+
+    pub fn as_ptr(&self) -> *const T {
+        self.as_slice().as_ptr()
+    }
+
+    pub fn as_mut_ptr(&mut self) -> *mut T {
+        self.as_mut_slice().as_mut_ptr()
+    }
+
+    pub fn clone<'b>(&'b mut self) -> CursorMut<'b, T> {
```

Comment:
Can you just `impl Clone`?

Reply:
We can't because `CursorMut` contains a `&mut [T]`, which can't be cloned. This method also works slightly different than the one provided by `Clone` in that it reborrows `self.data` rather than cloning it, which is why it also takes a `&mut self` instead of `&self`.

Reply:
I'm confused as to what the point of this method is then.  Is it just shortening the lifetime?

Reply:
Since `CursorMut` isn't `Copy`, the various arithmetic operations move the value. If you want to still use an older cursor after doing some math operations to create a new one, you need to explicitly clone the cursor. This happens in `selfguided_filter`:

```rust
// Create `AA` from `A` and then use `AA`.
let mut AA = A.clone() - REST_UNIT_STRIDE
for ... {
    AA += step * REST_UNIT_STRIDE;
}

// Go back to using `A`...
```

Reply:
I'm still not sure about the specifics why this is needed/only works this way, but it should be fine for now.